### PR TITLE
Drop retired MR entries on overlap deregistration

### LINKED ISSUE #615: Entries in MR cache are not being found
libfabric:gni:mr:__mr_cache_register():1367<trace> [34314:2] registering enter key=7fffaf0c6000:
10000
libfabric:gni:mr:print_tree():147<warn> [34314:2] contents of tree
libfabric:gni:mr:print_tree():153<warn> [34314:2]       key=7fffaf106000:10000
libfabric:gni:mr:__mr_cache_search_inuse():1021<info> [34314:2] could not find key in inuse, key
=7fffaf0c6000:10000
libfabric:gni:mr:__mr_cache_search_stale():1157<info> [34314:2] searching for stale entry, key=7
fffaf0c6000:10000
libfabric:gni:mr:_stale():1170<info> [34314:2] found a matching entry, found=7fffaf0b6000:160000
 key=7fffaf0c6000:10000
libfabric:gni:mr:__mr_cache_search_inuse():1021<info> [34314:2] could not find key in inuse, key
=7fffaf0b6000:160000
libfabric:gni:mr:__mr_cache_search_stale():1211<info> [34314:2] removing entry from stale and mi
grating to inuse, key=7fffaf0b6000:160000
libfabric:gni:mr:__mr_cache_register():1427<trace> [34314:2] registering return key=7fffaf0b6000
:160000

The existing entry in the inuse cache should have been merged with the new entry. It was not.

### PR DESCRIPTION
**Pull request by jswaro**
If a registration would be overlapped by a newly deregistered
memory registration, either the new stale registration or the old one
must be dropped. In the event that the newly deregistered region fully
subsumed several existing stale entries, those subsumed entries are
dropped immediately.

closes #615

### CODE REVIEW

**Comment by jswaro** [comment]
@chuckfossen @sungeunchoi @ztiffany @hppritcha please review when possible.

File: prov/gni/src/gnix_mr.c [hunk 1]
```diff
@@ -1100,21 +1185,9 @@ static int __mr_cache_search_inuse(
...
-				/* move the entry's children to the sibling tree
-				 * and decrement the reference count */
-				dlist_splice_tail(&(*entry)->children,
-						&found_entry->children);
-				__mr_cache_entry_put(cache, found_entry);
-			}
-		}
-		assert(dlist_empty(&tmp));
-		__mr_cache_entry_get(cache, *entry);
+	if (!dlist_empty(&retired_entries)) {
```
Comment (ztiffany) [inline prov/gni/src/gnix_mr.c:1185]: Isn't the list guaranteed to be non-empty?

Comment (jswaro) [inline prov/gni/src/gnix_mr.c:1185]: Probably.

**Comment by ztiffany** [comment]
:+1:

**Comment by sungeunchoi** [comment]
:+1:

**Pull request merged**